### PR TITLE
Add EgressRule to NetworkPolicy 

### LINKED ISSUE #50453: NetworkPolicy Egress with CIDRSelector and DNSSelector
/kind feature

**What happened**:
In 1.7 NetworkPolicy only supports ingress (although some cni providers like Calico claim to be able to also support egress).

No doubt that egress is already on the roadmap, but I'd like to state a requirement for a couple of selectors that we'll need in order for the egress functionality to be useful in enterprise environments.

1) CIDRSelector 

We need to be able to define egress rules from cluster pods to external services.  As these are external they lack both a namespace and any labels.  For something analogousness to traditional firewalling, it would be useful to define NetworkPolicy rules that allows egress to a number of network CIDRS or hosts. e.g.

```
---
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: egress-rule-for-logging
  namespace: backend
spec:
  podSelector:
    matchLabels:
      run: pod
  egress:
  - to:
    - CIDRSelector:
      - 192.168.0.0/16
      - 10.0.*
      - 192.169.0.12
    ports:
    - port: 80
```

2) DNSSelector

As more services are more dynamically IP'd (e.g. consul or an ELB etc.) it becomes hard to pinpoint an external service by CIDR without being too general.  In the same kind of vein as the ExternalCNAME on services it would be very useful to be able to create a egress rule that uses a DNS query to populate a list of IP Addresses in the rule (and to keep them updated by performing the dns check on a regular interval).  They could look something like this,,,

```
---
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: egress-rule-for-domaincontrollers
  namespace: backend
spec:
  podSelector:
    matchLabels:
      run: pod
  egress:
  - to:
    - DNSSelector:
      -  adfs.mycompanyforrest.ad.mycompany.com
    ports:
    - port: 636
```
In this example the dns query adfs.mycompanyforrest.ad.mycompany.com can be enumerated into a number of host ip addresses and egress allowed to this list only.  This would allow the IT team to add/remove servers at will without having to maintain a brittle egress rule.

**What you expected to happen**:
I'd like to see NetworkPolicy grow to include egress with new selectors that help ease the job of infrastructure developers defining rules for their infrastructure.

**How to reproduce it (as minimally and precisely as possible)**:
NA

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): 1.8?
- Cloud provider or hardware configuration**: AWS
- OS (e.g. from /etc/os-release): CoreOS
- Kernel (e.g. `uname -a`): Linux
- Install tools:
- Others:

/sig network

### PR DESCRIPTION
**Pull request by cmluciano**
**What this PR does / why we need it**:
Add EgressRule to NetworkPolicy

**Which issue this PR fixes**: fixes #50453

**Special notes for your reviewer**:
- Please take a look at the comments for the various types. I tried to mimic some of the language used in the Ingress comments, but I may have mangled some sentences.
- Let me know if I should add some test cases for validation. I have 2-3, and did not think it was necessary to replicate each case already covered in ingress.

**Release note**:
```
Add egress policies to NetworkPolicy
```

### CODE REVIEW

File: pkg/apis/extensions/v1beta1/defaults.go [hunk 1]
```diff
@@ -133,4 +133,14 @@ func SetDefaults_NetworkPolicy(obj *extensionsv1beta1.NetworkPolicy) {
 			}
 		}
 	}
+
+	if obj.Spec.PolicyTypes == nil || len(obj.Spec.PolicyTypes) == 0 {
+		// Default for undefined PolicyTypes with defined Egress rules
+		if obj.Spec.Egress != nil || len(obj.Spec.Egress) != 0 {
```
Comment (caseydavenport) [inline pkg/apis/extensions/v1beta1/defaults.go]: I know we discussed changing this yesterday, but I'm starting to doubt my memory.

```
spec:
  egress:
  - to: [ {...} ]
```

Should this policy affect ingress?  I'm thinking it probably shouldn't.

File: pkg/apis/extensions/v1beta1/defaults.go [hunk 2]
```diff
@@ -133,4 +133,14 @@ func SetDefaults_NetworkPolicy(obj *extensionsv1beta1.NetworkPolicy) {
...
+			obj.Spec.PolicyTypes = []extensionsv1beta1.PolicyType{extensionsv1beta1.PolicyTypeIngress, extensionsv1beta1.PolicyTypeEgress}
+			// Default for undefined PolicyTypes with undefined Egress rules
+		} else if obj.Spec.Egress == nil || len(obj.Spec.Egress) != 0 {
```
Comment (caseydavenport) [inline pkg/apis/extensions/v1beta1/defaults.go]: I don't think this clause is needed in the "else if" - `|| len(obj.Spec.Egress) != 0`

File: pkg/apis/extensions/v1beta1/defaults.go (hunk 1 above)
Comment (caseydavenport) [reply pkg/apis/extensions/v1beta1/defaults.go]: Looking back at the email thread where we designed this, I see:

```
  - [ "ingress", "egress" ] if the policy has both ingress: and egress:
    sections

  - [ "egress" ] if the policy has an egress: section but no ingress:
    section

  - [ "ingress"] if the policy has an ingress: section but no egress:
    section, or, for backward compatibility, if it has neither
    ingress: nor egress:.
```

So yeah, my memory was bad.

File: pkg/apis/networking/types.go [hunk 3]
```diff
@@ -36,6 +36,15 @@ type NetworkPolicy struct {
 	Spec NetworkPolicySpec
 }
 
+// Policy Type string describes the NetworkPolicy Rule type
```
Comment (caseydavenport) [inline pkg/apis/networking/types.go]: I think "NetworkPolicy Rule type" should just be "NetworkPolicy type"

File: pkg/apis/networking/types.go [hunk 4]
```diff
@@ -36,6 +36,15 @@ type NetworkPolicy struct {
...
+// This type is beta-level in 1.8
+type PolicyType string
+
+const (
+	PolicyTypeIngress PolicyType = "Ingress"
```
Comment (caseydavenport) [inline pkg/apis/networking/types.go]: I think each of these should have comments describing what they do.

e.g. "PolicyTypeIngress means a NetworkPolicy affects ingress traffic to selected pods"

File: pkg/apis/networking/types.go [hunk 5]
```diff
@@ -55,6 +64,24 @@ type NetworkPolicySpec struct {
...
+	// (and serves solely to ensure that the pods it selects does not allow any outgoing traffic.)
+	// This field is beta-level in 1.8
+	// +optional
+	Egress []NetworkPolicyEgressRule
+
+	// List of rule types that the NetworkPolicy relates to.
+	// Valid options are Ingress, Egress, or Ingress,Egress. If this field is not specified, it will
+	// default based on the existence of Ingress or Egress rules. PolicyTypes will always include Ingress
+	// in the list unless the field only contains Egress.
+	// This field is beta-level in 1.8
```
Comment (caseydavenport) [inline pkg/apis/networking/types.go]: Don't think we need the "in 1.8" part of this comment.

File: pkg/apis/networking/types.go [hunk 6]
```diff
@@ -55,6 +64,24 @@ type NetworkPolicySpec struct {
 	// solely to ensure that the pods it selects are isolated by default)
 	// +optional
 	Ingress []NetworkPolicyIngressRule
+
+	// List of egress rules to be applied to the selected pods. Outgoing traffic is
+	// allowed if there are no NetworkPolicies selecting the pod (and cluster policy
+	// otherwise allows the traffic), OR if the traffic matches at least one egress rule
+	// across all of the NetworkPolicy objects whose podSelector matches the pod. If
+	// this field is empty then this NetworkPolicy does not limit any outgoing traffic
```
Comment (caseydavenport) [inline pkg/apis/networking/types.go]: > // If this field is empty then this NetworkPolicy does not limit any outgoing traffic (and serves solely to ensure that the pods it selects does not allow any outgoing traffic.).

This sentence contradicts itself :)

File: pkg/apis/networking/validation/validation.go [hunk 7]
```diff
@@ -99,6 +152,8 @@ func ValidateNetworkPolicyUpdate(update, old *networking.NetworkPolicy) field.Er
 	return allErrs
 }
 
+// ValidatePolicyTypes validates the PolicyTypes field in a NetworkPolicySpec
```
Comment (caseydavenport) [inline pkg/apis/networking/validation/validation.go]: Is this missing a function body to go with the comment?

**Review by caseydavenport — CHANGES_REQUESTED** [review]

File: pkg/apis/extensions/v1beta1/defaults.go (hunk 2 above)
Comment (cmluciano) [reply pkg/apis/extensions/v1beta1/defaults.go]: typo, meant to say ==

File: pkg/apis/networking/validation/validation.go (hunk 7 above)
Comment (cmluciano) [reply pkg/apis/networking/validation/validation.go]: I think this is a mis-type. I was going to split this out in the future.

**Comment by cmluciano** [comment]
Working on addressing these changes locally.

**Comment by danwinship** [comment]
what happened to the old PR?

**Comment by cmluciano** [comment]
@danwinship had some rebase issues, pushed to old branch :P now

**Pull request closed without merging**